### PR TITLE
UI for managed DS

### PR DESCRIPTION
**Pull request by fontanierh**
- add section in DS for managed DS
- make settings and upsert RO for managed DS
- make inputs in RO mode look unfocusable
- prevent people from creating a DS name managed-*

todo:
- [x] log requests for early access
- [x] handle nango error
- [x] display `errored` status in DS list
- [ ] display last sync in settings
- [ ] ? display more detailed error in settings
- [ ] ? allow disconnecting (and/or delete ?) the DS from settings

### CODE REVIEW

File: front/pages/api/w/[wId]/data_sources/managed.ts [hunk 1]
```diff
@@ -107,7 +107,7 @@ async function handler(
 
       const connectorsRes = await ConnectorsAPI.createConnector(
         provider,
-        owner.id.toString(),
```
Comment (spolu) [inline front/pages/api/w/[wId]/data_sources/managed.ts:110]: No need for toString I think

File: front/pages/w/[wId]/ds/index.tsx [hunk 2]
```diff
@@ -54,8 +148,67 @@ export default function DataSourcesView({
...
+      });
+      if (res.ok) {
+        setManagedDataSourcesLocal((prev) =>
+          prev.map((ds) =>
+            ds.connectorProvider == provider
+              ? { ...ds, enabled: true, status: "succeeded" }
+              : ds
+          )
+        );
+      } else {
```
Comment (fontanierh) [inline front/pages/w/[wId]/ds/index.tsx:196]: @spolu  any idea what we should be doing here ? Not sure how I can really surface the issue to the user -- for now I was thinking of just assuming it's going to work

File: front/pages/w/[wId]/ds/index.tsx [hunk 3]
```diff
@@ -54,8 +148,67 @@ export default function DataSourcesView({
...
+    Record<ConnectorProvider, boolean | undefined>
+  >({} as Record<ConnectorProvider, boolean | undefined>);
+
+  const handleEnableManagedDataSource = async (provider: ConnectorProvider) => {
+    const nangoConnectorId =
+      provider == "slack" ? nangoSlackConnectorId : nangoNotionConnectorId;
+    const nango = new Nango({ publicKey: nangoPublicKey! });
+    const {
+      connectionId: nangoConnectionId,
+    }: { providerConfigKey: string; connectionId: string } = await nango.auth(
```
Comment (fontanierh) [inline front/pages/w/[wId]/ds/index.tsx:171]: @spolu  not sure what to do if nango fails here

File: front/pages/w/[wId]/ds/index.tsx [hunk 4]
```diff
@@ -168,6 +321,110 @@ export default function DataSourcesView({
...
+                          }
+                          onClick={
+                            canUseManagedDataSources
+                              ? () => {
+                                  handleEnableManagedDataSource(
+                                    ds.connectorProvider as ConnectorProvider
+                                  );
+                                }
+                              : () => {
+                                  logger.info(
```
Comment (fontanierh) [inline front/pages/w/[wId]/ds/index.tsx:390]: cc @spolu If you click on "Get early access" we simply log it. Is that good enough ?

File: front/pages/w/[wId]/ds/index.tsx [hunk 5]
```diff
@@ -168,6 +321,110 @@ export default function DataSourcesView({
...
+                                    },
+                                    "request_early_access_managed_data_source"
+                                  );
+                                }
+                          }
+                        >
+                          {!ds.isBuilt
+                            ? "Coming soon"
+                            : !canUseManagedDataSources
+                            ? "Get early access"
```
Comment (fontanierh) [inline front/pages/w/[wId]/ds/index.tsx:403]: @spolu @gabhubert  when a managed data source provider is already built on our end, but the user's workspace isn't whitelisted for managed DS, I replace the button with "Get early access" -- Is that OK ? Any wording improvement suggestion ?

File: front/pages/w/[wId]/ds/[name]/upsert.tsx [hunk 6]
```diff
@@ -205,8 +206,11 @@ export default function DataSourceUpsert({
                         id="document"
                         readOnly={readOnly}
                         className={classNames(
-                          "block w-full min-w-0 flex-1 rounded-md text-sm",
-                          "border-gray-300 focus:border-violet-500 focus:ring-violet-500"
+                          "block w-full min-w-0 flex-1 rounded-md border-gray-300 text-sm",
+                          "border-gray-300",
+                          readOnly
```
Comment (fontanierh) [inline front/pages/w/[wId]/ds/[name]/upsert.tsx:211]: cc @spolu , kind of unrelated, but for "Settings" and "Upsert", I removed the "focus" indicators on the inputs when they are readonly (felt a bit weird IMO to be able to focus them but not write)

**Comment by fontanierh** [comment]
@spolu I still have to work on displaying "last synced at" in the settings or something like that, but I can do it on a separate branch and we can merge that part when it looks good to you

File: front/pages/w/[wId]/ds/index.tsx (hunk 5 above)
Comment (spolu) [reply front/pages/w/[wId]/ds/index.tsx:403]: SGTM

Ideally if they click we invite to ping us at team@dust.tt. a window.alert is enough

File: front/pages/w/[wId]/ds/[name]/upsert.tsx (hunk 6 above)
Comment (spolu) [reply front/pages/w/[wId]/ds/[name]/upsert.tsx:211]: yep :+1:

File: front/pages/w/[wId]/ds/index.tsx [hunk 7]
```diff
@@ -1,24 +1,82 @@
...
+const MANAGED_DATA_SOURCES: ManagedDataSource[] = [
+  {
+    name: "Notion",
+    connectorProvider: "notion",
+    isBuilt: true,
+  },
+  {
+    name: "Slack",
+    connectorProvider: "slack",
+    isBuilt: true,
```
Comment (spolu) [inline front/pages/w/[wId]/ds/index.tsx:55]: let's set to false for now

Comment (fontanierh) [reply front/pages/w/[wId]/ds/index.tsx:55]: Ah yes

File: front/pages/w/[wId]/ds/index.tsx [hunk 8]
```diff
@@ -1,24 +1,82 @@
...
 export const getServerSideProps: GetServerSideProps<{
   user: UserType | null;
   owner: WorkspaceType;
   readOnly: boolean;
   dataSources: DataSourceType[];
+  managedDataSources: ManagedDataSource[];
+  canUseManagedDataSources: boolean;
   gaTrackingId: string;
+  nangoPublicKey: string;
+  nangoSlackConnectorId: string;
```
Comment (spolu) [inline front/pages/w/[wId]/ds/index.tsx:78]: can we pass the nango stuff as an object to avoid growing the list here?

**Comment by fontanierh** [comment]
Random screenshots dump:
Screenshot 2023-04-26 at 12 05 19
Screenshot 2023-04-26 at 12 05 29
Screenshot 2023-04-26 at 12 05 36
Screenshot 2023-04-26 at 12 05 42
5ea-4039-a24a-69cf99de33d5.png">

File: front/pages/w/[wId]/ds/index.tsx (hunk 2 above)
Comment (spolu) [reply front/pages/w/[wId]/ds/index.tsx:196]: also a window.alert will go a long way for the user

File: front/pages/w/[wId]/ds/index.tsx (hunk 4 above)
Comment (spolu) [reply front/pages/w/[wId]/ds/index.tsx:390]: as per below a small window,alert to invite to ping us

**Comment by spolu** [comment]
Left a few comments. Looks great. let me know when you added the window.alerts for approval

**Comment by fontanierh** [comment]
@spolu addressed the feedback !

File: connectors/src/api/syncStatus.ts [hunk 9]
```diff
@@ -36,7 +36,7 @@ const _getConnectorStatusAPIHandler = async (
       });
     }
     return res.status(200).send({
-      lastSyncStatus: connector.lastSyncStatus,
```
Comment (spolu) [inline connectors/src/api/syncStatus.ts:39]: probably etraneous?

File: front/pages/w/[wId]/ds/index.tsx [hunk 10]
```diff
@@ -387,6 +391,9 @@ export default function DataSourcesView({
                                   );
                                 }
                               : () => {
+                                  window.alert(
+                                    "Please reach out at team@dust.tt !"
```
Comment (spolu) [inline front/pages/w/[wId]/ds/index.tsx:395]: cc @gabhubert wording

**Review by spolu — CHANGES_REQUESTED** [review]
need to fix connector.lastSyncStatus ?
+ final wording for upgrade flow from @gabhubert

File: connectors/src/api/syncStatus.ts (hunk 9 above)
Comment (fontanierh) [reply connectors/src/api/syncStatus.ts:39]: Oups 😓

**Comment by fontanierh** [comment]
> need to fix connector.lastSyncStatus ?

changed that to test the error status and didn't mean to commit... Fixed

**Review by spolu — APPROVED** [review]

**Pull request merged**